mergify_engine/web/api/security.py

```diff
@@ -85,3 +88,50 @@ async def get_installation(
         )
     except exceptions.MergifyNotInstalled:
         raise fastapi.HTTPException(status_code=403)
+
+
+async def get_repository_context_with_queue_freeze_feat_check(
+    owner: github_types.GitHubLogin = fastapi.Path(  # noqa: B008
+        ..., description="The owner of the repository"
+    ),
+    repository: github_types.GitHubRepositoryName = fastapi.Path(  # noqa: B008
+        ..., description="The name of the repository"
+    ),
+    redis_cache: utils.RedisCache = fastapi.Depends(  # noqa: B008
+        redis.get_redis_cache
+    ),
+    redis_queue: utils.RedisQueue = fastapi.Depends(  # noqa: B008
+        redis.get_redis_queue
+    ),
+    installation_json: github_types.GitHubInstallation = fastapi.Depends(  # noqa: B008
+        get_installation
+    ),
+) -> context.Repository:
+
+    async with github.aget_client(installation_json) as client:
+        try:
+            # Check this token has access to this repository
+            repo = typing.cast(
+                github_types.GitHubRepository,
+                await client.item(f"/repos/{owner}/{repository}"),
+            )
+        except (http.HTTPNotFound, http.HTTPForbidden, http.HTTPUnauthorized):
+            raise fastapi.HTTPException(status_code=404)
+
+        sub = await subscription.Subscription.get_subscription(
+            redis_cache, installation_json["account"]["id"]
+        )
+
+        installation = context.Installation(
+            installation_json, sub, client, redis_cache, redis_queue
+        )
+
+        # Check this sub has access to queue_freeze feature
+        if installation.subscription.has_feature(subscription.Features.QUEUE_FREEZE):
+            return installation.get_repository_from_github_data(repo)
```

Comment:
I think you should `yield` the data instead of returning it. So the client will not be closed when the function return, but when the HTTP request finishes.